### ISSUE TITLE
Extra dependencies for Dependency-Management

### ISSUE THREAD

**Issue by adrianhurt** [michaelsauter/crane#170]
Now there are 3 different kinds of dependencies between containers: `link`, `volumes-from` and `net`. They allow crane to know the order it has to lift multiple containers. These 3 kind of dependencies correspond with the docker ones. In docker, they declare not only a dependency but also the way each container is affected by the other. However, when you try to manage multiple containers with crane, there are some situations where you need to establish an extra dependency without any other implication between them. I think a way to declare this extra dependency would allow us to tell crane that it has to lift the containers in a specific order.

# Automated Nginx
Imagine a nginx container that configures itself watching the containers instead of linking them manually ([jwilder/nginx-proxy](https://registry.hub.docker.com/u/jwilder/nginx-proxy/) is a good example). Then there's no need to define any `link` between each web container and nginx one. However, it's obvious there's a dependency between them. Each web container __needs__ the nginx to work. One possible way to declare this could be:
```
containers:
  web1:
    run: …
    needs: ["nginx"]
  web2:
    run: …
    needs: ["nginx"]

  nginx:
    image: jwilder/nginx-proxy
    run:
      publish: ["80:80"]
      …
```
Then, you can lift for example `web1` with `crane lift --cascade-dependencies=all web1`. Of course, you could solve this adding a group for each combination `webN-nginx` and lift the group, but I think it's interesting to declare this kind of dependency within the container's declaration.

# Restoring a Mysql backup
For example, imagine 3 containers: `db`, `db-restore-and-backup` and `web`. The container `db-restore-and-backup` needs to be started after `db` and it checks the content of the database. If there is no data, it restore the last backup it has. And of course it is in charge of backup the database everyday as well. 
```
containers:
  db:
    run:
      ...
  db-restore-and-backup:
    run:
      link: ["db:db"]
      volume: ["/db/backups:/db/backups"]
      ...
  web:
    run:
      link: ["db:db"]
      …
```
Then the `web` container needs to access to the database so we declare it with a `link: ["db:db"]`. However, when the containers are lifted the first time, the database is empty. It implies that `web` container should be started not only after `db` but also after `db-restore-and-backup`. I know there are different ways to solve that:
* Using sleeps... no way.
* Using extra "waiter" containers that uses `volumes-from` to workaround this issue. It lets you to define correctly the order of lifting each container, but I think adding extra containers and fake `volumes-from` declarations is not a very elegant way to solve this.
* Using active waiting loops to check when the database is ready to use. This needs to be done within the execution of the `web` container before starting the app. Well, I'm doing that now but I think it isn't as elegant as we could desire. Firstly, you must also define this extra dependency with fake `volumes-from` or using group declarations (I need to be sure `db-restore-and-backup` is always executed with `web`).

For that reasons, I think that the ability of declaring this extra dependency with the addition of the [support of container hooks feature](https://github.com/michaelsauter/crane/pull/159) would be the perfect solution for this kind of things. It could be resolve with something like this:
```
containers:
  db:
    ...
  db-restore-and-backup:
    ...
  web:
    run:
      link: ["db:db"]
      …
    needs: ["db-restore-and-backup"]

hooks:
  db-restore-and-backup:
    pre-start: for i in `seq 1 10`; do sleep 1 && docker exec db mysql -u root -prootpass database -e "show tables" && exit 0 || echo retrying...; done; exit 1
  web:
    pre-start:  for i in `seq 1 10`; do sleep 1 && docker exec db mysql -u root -prootpass database -e "describe one_of_my_tables" && exit 0 || echo retrying...; done; exit 1
```
With this, I'm sure that the order of execution will be `db` -> `db-restore-and-backup` -> `web`. And also that `db-restore-and-backup` will be started when the database is ready to use, and `web` when it has also the data initialized correctly, not before.

# Other options
Another alternative could be the use of `wait-for` and `wait-until`.
For the nginx example, we could use `wait-for`:
```
containers:
  web1:
    run: …
    wait-for: ["nginx"]
  web2:
    run: …
    wait-for: ["nginx"]

  nginx:
    image: jwilder/nginx-proxy
    run:
      publish: ["80:80"]
      …
```
And for the mysql example, we could use `wait-until`:
```
containers:
  db:
    ...
  db-restore-and-backup:
    run:
      link: ["db:db"]
      …
    wait-until:
      db: mysql -u root -prootpass database -e "show tables"
  web:
    run:
      link: ["db:db"]
      …
    wait-for: ["db-restore-and-backup"]
    wait-until:
      db: mysql -u root -prootpass database -e "describe one_of_my_tables"
```
Where `wait-until` would wait until the specified command would return a _0 status_ (with an active waiting loop and `docker exec` functionality).

**Comment by adrianhurt** [comment]
# Another option
Other possibility could be ability for each container to define a way to know if it's ready.
```
containers:
  db:
    …
    ready-when: mysql -u root -prootpass database -e "show tables"
  db-restore-and-backup:
    run:
      link: ["db:db"]
      …
    wait-for: ["db"]
    ready-when: mysql -u root -prootpass database -e "describe one_of_my_tables"
  web:
    run:
      link: ["db:db"]
      …
    wait-for: ["db-restore-and-backup"]
    ready-when: curl localhost:9000 | grep '<h1>Home</h1>'
```
Where the command defined by `ready-when` is executed by crane using `docker exec`, and the `wait-for` would wait until the corresponding `ready-when` of the target container is satisfied.

These are only options fore new features. But I think it could enhance crane functionality.

**Comment by adrianhurt** [comment]
I've decided to split this issue in two. I've just opened another one (#171) to discuss about the ability to declare the readiness of a container. And this issue is only to discuss the ability to add an extra dependency. So then you can ignore the `wait-for` and `wait-until` examples. Let's focus only in that extra dependency `needs` (or other name, of course).

**Comment by dreamcat4** [comment]
+1 For any way to implement this feature (user-controlled dependancies). Rather than fallback to Alphabetical (the current).

P.S. A simple hack until then is to prefix your container names with `a,b,c,d,e,f,g,…etc` to force the startup order. E.g.

```yaml
containers:
  a_my_dependant_service_1:
    image: elasticsearch
  b_my_dependant_service_2:
    image: mysql
  c_my_web_application:
    image: rails/rails
```

Then your containers will always fail over to starting in the order `a,b,c`. Sorry it's an undocumented hack though. It would be better to make as a proper feature ~ I have no objection to your proposal to use `needs:` key however there may be other suitable key names, e.g. `depends:` comes to mind as an obvious one.

Writing `needs:` suggests general and multiple possible types (kinds) of dependancy. e.g. including not just other containers to start up. Wheras `depends:` is un-ambiguous to containers. My other caveat (in other issue #171) also stands here too:

We might as well be thorough about it and check if some similar feature is already in the works for the official docker project. Before taking that name. To avoid possible future conflicts / duplication / and confusion over official docker features.

**Comment by adrianhurt** [comment]
Totally agree with using `depends:` instead of `needs:`. And as I answered you in the other issue, I think it's not really part of Docker (CLI). It only cares about giving you the tools to run individual containers. It delegates to the user to decide the specific order of each container. It would be a Compose issue, not Docker. And Crane is a substitute of Compose, so...

**Comment by bjaglin** [comment]
I definitely see the need for materializing dependencies that are not captured by the current dependency constraints, even though it's an advanced use-case IMHO.

If we want to go that way, how about going with systemd semantics (`Requires`/`After`/`Before`) ? Distinguishing needs from ordering would allow what https://github.com/michaelsauter/crane/pull/149 was asking for for these non-materialized constraints.

**Comment by dreamcat4** [comment]
+1 `requires:`, `after:`, `before`

**Comment by dreamcat4** [comment]
+1 `requires:`, `after:`, `before:`

**Comment by adrianhurt** [comment]
It sounds me perfect!

**Comment by dreamcat4** [comment]
It's a shame that docker itself does not provide a built-in mechanism for determining dependancies. Because then it would be possible to use docker's restart policies to bring back up in correct order the previously running containers:

http://serverfault.com/a/649835

**Comment by michaelsauter** [comment]
Tentatively :+1:

**Comment by mishak87** [comment]
TL;DR I am for it, would like to help with it after CI for this project is set up. Require and wait would be nice to have. Still such level of orchestration is better suited for systemd or similar systems and once you need them you don't need crane.

@dreamcat4 Docker really shines in systems like CoreOS where advanced logic is handled by systemd unit files.

@adrianhurt Some of your examples have docker-ish ways of being done.

Nginx can be listening on docker socket for new containers like in https://github.com/jwilder/nginx-proxy or can be configured via etcd/zoo or ambassador pattern (one app tells another that is should update its config or creates bridges that connect them).

MySQL container should have logic for restoring from backup included, usually slave/master detection and signalling that it is up to date, such logic is useful for bringing up new nodes so its well worth it. In production you will have at least master and slave setup and backup will be done by shutting down slave, backing up its volume and starting it back up.

**Comment by adrianhurt** [comment]
Hi again @michaelsauter. Is it on the roadmap? `requires:`, `after:` and `before:` sounds great but maybe with only `requires:` is enough for the moment, and simpler to implement.

**Comment by michaelsauter** [comment]
There is no real "roadmap", it's more a case of what's required by people is being added. Personally, I don't have a need for this feature and think the workaround of just adding a link is good enough for most cases. I'm willing to have a look at a PR though and include it if it turns out to be quite straight-forward.

**Comment by michaelsauter** [comment]
I was just looking into the new networking feature of Docker. Given that we definitely have a need for requires now since links do not work in this case, but we might want to ensure some other containers are started apart from the target. I'm going to add network support and `requires` soon-ish.

**Comment by adrianhurt** [comment]
Great!

**Comment by michaelsauter** [comment]
This will land in master soon as port of #248.
